test/e2e/http_header_buffer_test.go

```diff
@@ -0,0 +1,310 @@
+		defer func() {
+			if err := readCloser.Close(); err != nil {
+				t.Errorf("failed to close reader for pod %s: %v", clientPodValidRequest.Name, err)
+			}
+		}()
+		for scanner.Scan() {
+			line := scanner.Text()
+			if idx := strings.Index(line, ":"); idx != -1 {
+				// Ensure that we see the test header name in
+				// the echo response. We don't need to verify the
+				// contents of the header since HAProxy would return
+				// 400 if the header was too large.
+				headerName := line[:idx]
+				if headerName == testHeaderName {
+					return true, nil
+				}
+			}
+		}
+		return false, nil
+	})
+	if pollErr != nil {
+		pod := &corev1.Pod{}
+		podName := types.NamespacedName{
+			Namespace: clientPodValidRequest.Namespace,
+			Name:      clientPodValidRequest.Name,
+		}
+		if err := kclient.Get(context.TODO(), podName, pod); err != nil {
+			t.Errorf("failed to get pod %s: %v", clientPodValidRequest.Name, err)
+		}
+
+		logs, err := cl.CoreV1().Pods(clientPodValidRequest.Namespace).GetLogs(clientPodValidRequest.Name, &corev1.PodLogOptions{
+			Container: "curl",
+			Follow:    false,
+		}).DoRaw(context.TODO())
+		if err != nil {
+			t.Errorf("failed to get logs from pod %s: %v", clientPodValidRequest.Name, err)
+		}
+
+		output := fmt.Sprintf("failed to observe the expected output: %v\nclient pod spec: %#v\nclient pod logs:\n%s", pollErr, pod, logs)
+		output = strings.Replace(output, random, fmt.Sprintf("<random data of len %d btyes snipped>", length), 0)
+		t.Fatal(output)
+	}
+
+	if err := kclient.Get(context.TODO(), types.NamespacedName{Name: ic.Name, Namespace: ic.Namespace}, ic); err != nil {
+		t.Errorf("failed to get ingresscontroller %s: %v", ic.Name, err)
+	}
+
+	// Get the name of the current router pod for the test ingress controller
+	podList := &corev1.PodList{}
+	labels := map[string]string{
+		controller.ControllerDeploymentLabel: "header-buffer-size",
+	}
+	if err := kclient.List(context.TODO(), podList, client.InNamespace(deployment.Namespace), client.MatchingLabels(labels)); err != nil {
+		t.Errorf("failed to list pods for ingress controllers %s: %v", ic.Name, err)
+	}
+
+	if len(podList.Items) != 1 {
+		t.Errorf("expected ingress controller %s to have exactly 1 router pod, but it has %d", ic.Name, len(podList.Items))
+	}
+
+	oldRouterPodName := podList.Items[0].Name
+
+	// Mutate the ic to use header buffer values that are 1/2 the default.
+	ic.Spec.HTTPHeaderBuffer = operatorv1.IngressControllerHTTPHeaderBuffer{
+		HeaderBufferBytes:           16384,
+		HeaderBufferMaxRewriteBytes: 4096,
+	}
+
+	if err := kclient.Update(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to update ingresscontroller %s: %v", icName, err)
+	}
+
+	if err := waitForIngressControllerCondition(t, kclient, 5*time.Minute, icName, conditions...); err != nil {
+		t.Fatalf("failed to observe expected conditions: %v", err)
+	}
+
+	// Wait for the new router pod for the updated ingresscontroller to become ready.
```

Comment:
Hm, maybe it would be simpler. Mind if I improve on this e2e test in a follow up? Wondering if my current approach might actually shave some time off of the test (But I guess we aren't super worried about how long these tests take).

Reply:
Follow-up is fine.